libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -286,16 +294,67 @@ string IRGenerator::constructorCode(ContractDefinition const& _contract)
 
 			<constructorName>(<params>)
 		)X");
-		t("object", creationObjectName(_contract));
-		t("allocate", m_utils.allocationFunction());
-		t("assignToParams", paramVars == 0 ? "" : "let " + suffixedVariableNameList("param_", 0, paramVars) + " := ");
-		t("params", suffixedVariableNameList("param_", 0, paramVars));
-		t("abiDecode", abiFunctions.tupleDecoder(constructor->functionType(false)->parameterTypes(), true));
-		t("constructorName", m_context.functionName(*constructor));
-
-		out << t.render();
+	t("object", creationObjectName(dynamic_cast<ContractDefinition const&>(*_constructor.scope())));
+	t("allocate", m_utils.allocationFunction());
+	t("assignToParams", params.empty() ? "" : "let " + params + " := ");
+	t("abiDecode", abiFunctions.tupleDecoder(_constructor.functionType(false)->parameterTypes(), true));
+	t("constructorName", m_context.functionName(_constructor));
+	t("params", params);
+
+	return t.render();
+}
+
+std::string IRGenerator::addBaseContractConstructor(FunctionDefinition const& _baseConstructor)
+{
+	FunctionType constructorType(_baseConstructor);
+	string params;
+	IRGeneratorForStatements generator(m_context, m_utils);
+	if (!constructorType.parameterTypes().empty())
+	{
+		solAssert(m_baseArguments, "");
+		solAssert(m_baseArguments->count(&_baseConstructor), "");
+		std::vector<ASTPointer<Expression>> const* arguments = nullptr;
+		ASTNode const* baseArgumentNode = m_baseArguments->at(&_baseConstructor);
```

Comment:
I think we have to talk about a big conceptual question here first:
Should the arguments be evaluated at the time the base constructor is called or at the time where the constructor of the contract that contains the arguments is visited? I think the cleaner approach would be the latter.

Example:

The proper way to initialize would be:
 - initialize all state variables from base to derived
 - evaluate base constructor arguments from derived to base
 - execute constructor bodies from base to derived

So in the example below, the order should be:
 - initialize `x=7`
 - evaluate `A(++x)`
 - evaluate `B(++x)`
 - evaluate A's constructor code
 - evaluate B's constructor code
 - evaluate C's constructor code

```
contract A {
  uint public x = 7;
  uint public a;
  constructor(uint _a) public {
    a = _a;
  }
}
abstract contract B is A {
  uint public b;
  constructor(uint _b) public {
    b = _b;
  }
}
contract C is A, B {
  constructor() public A(++x) B(++x) {
  }
}
```

I know that this is not what the evmasm compiler is doing, but I think this is how it should be.

Reply:
We should force the order of base constructor calls to follow the linearization then.
But I'm not entirely sure about the order. It also does make sense to finalize the base before moving to the derived, i.e.
- initialize x = 7
- evaluate A(++x)
- evaluate A's contructor code
- evaluate B(++x)
- evaluate B's constructor code
- evaluate C's constructor code

If I'd initialize ``uint public b = a;`` in ``B`` I'd rather expect that to be the *post*-constructor value of a, not the pre-constructor value...

Reply:
Yeah I think we have an unsolvable circular dependency here. If base constructor arguments reference base contract state variables - are they already initialized or not?

Reply:
An argument for "finalizing" the base contract early is that functions in bases will return potentially weird data if used for initialization in derived contracts...
```
contract A {
  uint minimum;
  constructor(uint _m) public {
    minimum = _m;
  }
  uint getMinimum() public { return minimum; }
}
contract B is A {
  constructor(uint maximum) public { /* ... */  }
}
contract C is B {
  uint public twiceMinimum = 2 * getMinimum(); // I'd argue even for these the constructor of A should already have been called.
  constructor() A(5) B(getMinimum() * 10) public {}
}
```
On the other hand: when calling a function of A in the constructor arguments of A itself, it's clear that I can't yet expect A to be "already completely constructed".
But I may be biased by the C++ way to do things (see https://en.cppreference.com/w/cpp/language/initializer_list#Initialization_order)

Reply:
Ok, we want to generally do this just like C++ does, but that's slightly different from what I said earlier, so what we need to do is:

- evaluate all constructor arguments (we have to decide the order, though, see below)
- Construct contracts from most base to most derived. For each of them:
  - first initialize state variables
  - then execute constructor

Regarding the order in which to evaluate constructor arguments:
C++ generally evaluates the arguments to most derived constructors first and then the more base ones. We should generally reproduce that, but solidity is different from C++ here, since you can't specify arguments for base-of-base-constructors in C++. So for solidity what we should do is to first look at the base constructor calls in the most derived contract and evaluate all of them in the order they appear in the source (or in the reverse order, so far we said we're fine with either way). Then move to the next base constructor and evaluate the arguments there and so on.